components/tidb_query_datatype/src/codec/mysql/time/interval.rs

```diff
@@ -0,0 +1,1774 @@
+            ),
+            (
+                "-2.1234567",
+                Second,
+                Duration::from_nanos(-2 * NANOS_PER_SEC - 123456 * NANOS_PER_MICRO, 6),
+            ),
+            (
+                "2.99",
+                Second,
+                Duration::from_nanos(2 * NANOS_PER_SEC + 990000 * NANOS_PER_MICRO, 2),
+            ),
+            (
+                "-2.50000",
+                Second,
+                Duration::from_nanos(-2 * NANOS_PER_SEC - 500000 * NANOS_PER_MICRO, 5),
+            ),
+            ("99", Minute, Duration::from_nanos(99 * NANOS_PER_MIN, 0)),
+            ("-99", Minute, Duration::from_nanos(-99 * NANOS_PER_MIN, 0)),
+            ("30", Day, Duration::from_nanos(30 * NANOS_PER_DAY, 0)),
+            ("-30", Day, Duration::from_nanos(-30 * NANOS_PER_DAY, 0)),
+            ("2", Week, Duration::from_nanos(2 * NANOS_PER_DAY * 7, 0)),
+            ("-2", Week, Duration::from_nanos(-2 * NANOS_PER_DAY * 7, 0)),
+            ("1", Month, Duration::from_nanos(30 * NANOS_PER_DAY, 0)),
+            ("-1", Month, Duration::from_nanos(-30 * NANOS_PER_DAY, 0)),
+            (
+                "29 12:23:36.1234",
+                DayMicrosecond,
+                Duration::from_nanos(
+                    29 * NANOS_PER_DAY
+                        + 12 * NANOS_PER_HOUR
+                        + 23 * NANOS_PER_MIN
+                        + 36 * NANOS_PER_SEC
+                        + 123400 * NANOS_PER_MICRO,
+                    6,
+                ),
+            ),
+            (
+                "-29 12:23:36.1234",
+                DayMicrosecond,
+                Duration::from_nanos(
+                    -29 * NANOS_PER_DAY
+                        - 12 * NANOS_PER_HOUR
+                        - 23 * NANOS_PER_MIN
+                        - 36 * NANOS_PER_SEC
+                        - 123400 * NANOS_PER_MICRO,
+                    6,
+                ),
+            ),
+        ];
+        let mut ctx = EvalContext::default();
+        for (input, unit, expected) in cases {
+            let result = Interval::extract_duration(&mut ctx, &unit, input).unwrap();
+            assert_eq!(
+                result,
+                expected.unwrap(),
+                "Failed for input: {}, unit: {:?}",
+                input,
+                unit
+            );
+        }
+        let err_cases = vec![
+            ("2.500000", Minute),
+            ("-2.50000", Minute),
+            ("99.9", Hour),
+            ("-99.4", Hour),
+            ("35", Day),
+            ("-35", Day),
+            ("2", Month),
+            ("-2", Month),
+            ("99", Quarter),
+            ("-99", Quarter),
+            ("99", Year),
+            ("-99", Year),
+            ("-34 23:59:59.1234", DayMicrosecond),
+        ];
+        for (input, unit) in err_cases {
+            let result = Interval::extract_duration(&mut ctx, &unit, input);
```

Comment:
You can use proptest[1] for testing parsing methods, 

\[1\]: https://proptest-rs.github.io/proptest/proptest/getting-started.html

Reply:
I will try it next time due to lack of time.